### PR TITLE
feat(gcli): embed programs in commands

### LINKED ISSUE #3587: Introduce features `embed` for gcli
### Problem to Solve

After #3573, we have introduced a interface that constructing CLI for programs, once we present a CLI based on program, we actually can get the compiled WASM of the program directly from the target folder

### Possible Solution

when `embed` feature enabled, skipping parsing program path from command line arguments but initializing related commands with the built binaries

### Notes

_No response_

### LINKED ISSUE #3592: Override WASM binary in embedded CLI
### Problem to Solve

for command `upload`, https://github.com/gear-tech/gear/pull/3589 currently just overrides the path of the built program for the compatible with the previous API, we should override the binary in the future in case of users renaming the path of the binaries.

### Possible Solution

mb adding a new field `overided_code` in command `Upload`, or wait for the implementation of **the commands for handle inputs**, and do it together in the future

### Notes

_No response_

### PR DESCRIPTION
**Pull request by clearloop**
Resolves #3587 
Resolves #3592 
Ref #3588

- [x] introduce tools for looking up wasm binaries
- [x] introduce features `embed`
- [x] nest binaries for related command if feature `embed` is enabled
- [x] example
- [x] tests

@gear-tech/dev

### CODE REVIEW

**Review by breathx — APPROVED** [review]

File: examples/messager/Cargo.toml [hunk 1]
```diff
@@ -5,8 +5,14 @@ authors.workspace = true
 license.workspace = true
 edition.workspace = true
 
+[[bin]]
+name = "demo_messager"
```
Comment (shamilsan) [inline examples/messager/Cargo.toml:9]: Maybe `messenger`?

https://english.stackexchange.com/questions/211365/why-is-messenger-the-term-instead-of-messager

File: examples/messager/Cargo.toml [hunk 2]
```diff
@@ -5,8 +5,14 @@ authors.workspace = true
...
+path = "src/bin/demo_messager.rs"
+required-features = ["gcli"]
+
 [dependencies]
 gstd.workspace = true
+gcli = { workspace = true, optional = true, features = [ "embed" ] }
```
Comment (shamilsan) [inline examples/messager/Cargo.toml:15]:
```suggestion
gcli = { workspace = true, optional = true, features = ["embed"] }
```

File: gcli/src/cmd/program.rs [hunk 3]
```diff
@@ -60,6 +64,15 @@ pub enum Program {
 }
 
 impl Program {
+    /// Override metadata binary.
+    pub fn override_meta(&self, meta: Vec<u8>) -> Self {
```
Comment (shamilsan) [inline gcli/src/cmd/program.rs:68]: Maybe we can reflect that we clone `self` in the function name.
```suggestion
    pub fn clone_with_meta_override(&self, meta: Vec<u8>) -> Self {
```

File: gcli/src/embed.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            } else {
+                "release"
+            },
+            ancestors.next()?.to_str()?,
+        ];
+
+        let [bin, stem] = [
+            PathBuf::from(format!("{target}/wasm32-unknown-unknown/{profile}")),
+            PathBuf::from(name.replace('-', "_")),
+        ];
```
Comment (shamilsan) [inline gcli/src/embed.rs:80]: Why array with destructuring?
```suggestion
        let bin = PathBuf::from(format!("{target}/wasm32-unknown-unknown/{profile}"));
        let stem = PathBuf::from(name.replace('-', "_"));
```

File: gcli/src/embed.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+        let [opt, meta] = [
+            fs::read(bin.join(stem.with_extension("wasm"))).ok()?,
+            fs::read(bin.join(stem.with_extension("meta.wasm"))).unwrap_or_default(),
+        ];
+
+        Some(Self { opt, meta })
```
Comment (shamilsan) [inline gcli/src/embed.rs:87]: Also,
```suggestion
        Some(Self {
            opt: fs::read(bin.join(stem.with_extension("wasm"))).ok()?,
            meta: fs::read(bin.join(stem.with_extension("meta.wasm"))).unwrap_or_default(),
        })
```

**Review by shamilsan — APPROVED** [review]

File: gcli/src/cmd/program.rs (hunk 3 above)
Comment (clearloop) [reply gcli/src/cmd/program.rs:68]: makes sense, and this method should be used by users in common cases

File: examples/messager/Cargo.toml (hunk 1 above)
Comment (clearloop) [reply examples/messager/Cargo.toml:9]: https://github.com/gear-tech/gear/issues/3621 will do this in another PR since this related to workspace changes

**Pull request merged**